### ISSUE TITLE
OSNet can't reproduce the result in paper on CUHK03 and MSMT17 dataset

### ISSUE THREAD

**Issue by valuefish** [KaiyangZhou/deep-person-reid#279]
I have trained OSNet on Market1501、DukeMTMC-reID、CHUK03 and MSMT17 with referring to the paper regarding the hyperparameter setting https://arxiv.org/abs/1910.06827. And I got **the same result** as paper on Market1501 and DukeMTMC-reID dataset. But when it comes to CUHK03 and MSMT17 dataset, I **can't reproduce the result** (same hyperparameters as Market1501).

The results are listing below:

| Dataset | Expected Rank-1 | Expected mAP | Reprodued Rank-1 | Reproduced mAP  | Train Mode   |
| ------- | --------------- | ------------ | ---------------- | --------------- | ------------ |
| CUHK03  | 72.3            | 67.8         | 67.6 **(-4.7)**  | 63.7 **(-4.1)** | fine-tune    |
| CUHK03  | 57.1            | 54.2         | 51.9 **(-5.2)**  | 50.2 **(-4)**   | scratch |
| MSMT17  | 78.7            | 52.9         | 78.2             | 50.2 **(-2.7)** | fine-tune    |
| MSMT17  | 71              | 43.4         | 69.3             | 39.9 **(-3.5)** | scratch |

For CUHK03, I used the 767/700 split version. For MSMT17, I used the V1 version.

**Comment by KaiyangZhou** [comment]
please provide
- the full commands (config, args, etc.)
- version of torchreid (which commit)
- system info (pytorch version ...)

**Comment by KaiyangZhou** [comment]
also, please do a search in the issue history where you might find your answer

**Comment by valuefish** [comment]
Thank you for your prompt reply! I have done the search for issue history and got no answer.

The supplementary information is listing below:

#### CUHK03 config
MSMT17 configs are the same as CUHK03 except for dataset name.

##### ImageNet fine-tune
```yaml
data:
  type: 'image'
  sources: ['cuhk03']
  targets: ['cuhk03']
  height: 256
  width: 128
  combineall: False
  transforms: ['random_flip', 'random_erase']
  save_dir: 'log/osnet_x1_0_cuhk03_softmax_cosinelr'

loss:
  name: 'softmax'
  softmax:
    label_smooth: True

train:
  optim: 'amsgrad'
  lr: 0.0015
  max_epoch: 250
  batch_size: 64
  fixbase_epoch: 10
  open_layers: ['classifier']
  lr_scheduler: 'cosine'

test:
  batch_size: 300
  dist_metric: 'euclidean'
  normalize_feature: False
  evaluate: False
  eval_freq: -1
  rerank: False
  visactmap: False
```

##### Train from scratch
```yaml
model:
  name: 'osnet_x1_0'
  pretrained: False

data:
  type: 'image'
  sources: ['cuhk03']
  targets: ['cuhk03']
  height: 256
  width: 128
  combineall: False
  #  transforms: ['random_flip']
  transforms: ['random_flip', 'random_crop', 'random_patch']
  save_dir: 'log/osnet_x1_0_cuhk03_softmax_sgdlr'

loss:
  name: 'softmax'
  softmax:
    label_smooth: True

train:
  optim: 'sgd'
  lr: 0.065
  max_epoch: 350
  batch_size: 64
  fixbase_epoch: 10
  open_layers: ['classifier']
  lr_scheduler: 'multi_step'
  stepsize: [150, 225, 300]

test:
  batch_size: 300
  dist_metric: 'euclidean'
  normalize_feature: False
  evaluate: False
  eval_freq: -1
  rerank: False
  visactmap: False
```

#### Train command
```shell
python scripts/main.py --config-file configs/im_osnet_x1_0_softmax_256x128_amsgrad_cosine_cuhk03.yaml --root /home/datasets/
```

#### Version of torchreid
https://github.com/KaiyangZhou/deep-person-reid/commit/7ff428608ae98e3d2fdf98dfd22b101352f9fb1d

#### System info
```shell
PyTorch version: 1.3.1+cu100
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: CentOS Linux release 7.6.1810 (Core)
GCC version: (GCC) 5.3.1 20160406 (Red Hat 5.3.1-6)
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: Tesla V100-PCIE-16GB
GPU 1: Tesla V100-PCIE-16GB
GPU 2: Tesla V100-PCIE-16GB
GPU 3: Tesla V100-PCIE-16GB

Nvidia driver version: 410.104
cuDNN version: /usr/lib64/libcudnn.so.7.6.3
```

#### Full train log
CUHK03 fine-tune: [osnet_x1_0_cuhk03_softmax_cosinelr.txt](https://github.com/KaiyangZhou/deep-person-reid/files/3993375/osnet_x1_0_cuhk03_softmax_cosinelr.txt)
CUHK03 scratch: [osnet_x1_0_cuhk03_softmax_sgdlr.txt](https://github.com/KaiyangZhou/deep-person-reid/files/3993380/osnet_x1_0_cuhk03_softmax_sgdlr.txt)
MSMT17 fine-tune: [osnet_x1_0_msmt17_softmax_cosinelr.txt](https://github.com/KaiyangZhou/deep-person-reid/files/3993388/osnet_x1_0_msmt17_softmax_cosinelr.txt)
MSMT17 scratch: [osnet_x1_0_msmt17_softmax_sgdlr.txt](https://github.com/KaiyangZhou/deep-person-reid/files/3993393/osnet_x1_0_msmt17_softmax_sgdlr.txt)

**Comment by KaiyangZhou** [comment]
at a glance, the config files look good but not exactly the same (not sure if you have done sth else)

in the finetuning config, set `dist_metric=cosine` or equivalently, use `normalize_feature=true`, which may bring some improvement to the mAP (also make sure you have the right imagenet weights that are correctly loaded)

in the train-from-scratch file, set `fixbase_epoch=0` (all the weights are randomly initialized so there is no need to fix the base network)

and try with *different* seeds

if you are building your model on top of OSNet, it's better to compare the results obtained in the same environment (also quote the results in the original paper for reference)

ps. I like your gpus

**Comment by valuefish** [comment]
Thanks for your advice, I will try that.

ps. hope my GPUs can bring me excellent model like OSNet 😁.

**Comment by valuefish** [comment]
I have tried it in CUHK03 according to your suggestion, strangely, I still can't reproduce the result of the paper.

For fine-tuning,

| Experiment                        | Rank-1          | mAP             |
| --------------------------------- | --------------- | --------------- |
| Expected Result                   | 72.3            | 67.8            |
| `seed=2`, `dist_metric=euclidean` | 67.6 **(-4.7)** | 63.6 **(-4.2)** |
| `seed=2`, `dist_metric=cosine`    | 69.1 **(-3.2)** | 65.6 **(-2.2)** |
| `seed=3`, `dist_metric=cosine`    | 67.5 **(-4.8)** | 63.7 **(-4.1)** |
| `seed=4`, `dist_metric=cosine`    | 69.1 **(-3.2)** | 65.7 **(-2.1)** |

I also tried `normalize_feature=true` which result same as the form above.

For training from scratch,

| Experiment                                           | Rank-1          | mAP             |
| ---------------------------------------------------- | --------------- | --------------- |
| Expected Result                                      | 57.1            | 54.2            |
| `seed=1`, `dist_metric=euclidean`, `fixbase_epoch=0` | 53.1 **(-4.0)** | 50.8 **(-3.4)** |
| `seed=1`, `dist_metric=cosine`, `fixbase_epoch=0`    | 55.1 **(-2.0)** | 53.5 **(-0.7)** |
| `seed=2`, `dist_metric=cosine`, `fixbase_epoch=0`    | 54.6 **(-2.5)** | 53.5 **(-0.7)** |

Maybe there is still some difference in our config, can you share your config file used in OSNet paper with me?

**Comment by KaiyangZhou** [comment]
not sure why your results are low on cuhk03 (so msmt17's results are normal?)

I just tried [this config](https://github.com/KaiyangZhou/deep-person-reid/blob/master/configs/im_osnet_x1_0_softmax_256x128_amsgrad_cosine.yaml) and got
```
##### Evaluating cuhk03 (source) #####
Extracting features from query set ...
Done, obtained 1400-by-512 matrix
Extracting features from gallery set ...
Done, obtained 5332-by-512 matrix
Speed: 0.3683 sec/batch
Computing distance matrix with metric=euclidean ...
Computing CMC and mAP ...
** Results **
mAP: 66.6%
CMC curve
Rank-1  : 70.9%
Rank-5  : 84.3%
Rank-10 : 88.9%
Rank-20 : 92.7%
```
not exactly the same but quite close

below is the printed config
```
Show configuration
adam:
  beta1: 0.9
  beta2: 0.999
cuhk03:
  classic_split: False
  labeled_images: False
  use_metric_cuhk03: False
data:
  combineall: False
  height: 256
  norm_mean: [0.485, 0.456, 0.406]
  norm_std: [0.229, 0.224, 0.225]
  root: /data/reid
  save_dir: log/osnet/osnet_x1_0_cuhk03
  sources: ['cuhk03']
  split_id: 0
  targets: ['cuhk03']
  transforms: ['random_flip', 'random_erase']
  type: image
  width: 128
  workers: 4
loss:
  name: softmax
  softmax:
    label_smooth: True
  triplet:
    margin: 0.3
    weight_t: 1.0
    weight_x: 0.0
market1501:
  use_500k_distractors: False
model:
  load_weights: 
  name: osnet_x1_0
  pretrained: True
  resume: 
rmsprop:
  alpha: 0.99
sampler:
  num_instances: 4
  train_sampler: RandomSampler
sgd:
  dampening: 0.0
  momentum: 0.9
  nesterov: False
test:
  batch_size: 300
  dist_metric: euclidean
  eval_freq: -1
  evaluate: False
  normalize_feature: False
  ranks: [1, 5, 10, 20]
  rerank: False
  start_eval: 0
  visactmap: False
  visrank: False
  visrank_topk: 10
train:
  base_lr_mult: 0.1
  batch_size: 64
  fixbase_epoch: 10
  gamma: 0.1
  lr: 0.0015
  lr_scheduler: cosine
  max_epoch: 250
  new_layers: ['classifier']
  open_layers: ['classifier']
  optim: amsgrad
  print_freq: 20
  seed: 1
  staged_lr: False
  start_epoch: 0
  stepsize: [20]
  weight_decay: 0.0005
use_gpu: True
video:
  pooling_method: avg
  sample_method: evenly
  seq_len: 15
```
